File: 29_divide_two_integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        if dividend == 0:
            return 0

        import math
        up = int(math.fabs(dividend))
        down = int(math.fabs(divisor))
        result = 0
        while up >= down:
            base = down
            count = 1

            while up >= (base << 1):
                base <<= 1
                count <<= 1

            result += count
            up -= base

        if dividend ^ divisor < 0:
            result = -result

        if result < -(1 << 31) or result > ((1 << 31) - 1):
            result = (1 << 31) - 1

        return result


int1 = -2147483648
int2 = -1
solution = Solution()
print(solution.divide(int1, int2))

File: 29_divide_two_integers.go
package main

func divide(dividend int, divisor int) int {
	if dividend == -(1 << 31) {
		if divisor == -1 {
			return 1<<31 - 1
		}
		if divisor == 1 {
			return -(1 << 31)
		}
	}
	positive := false
	if dividend > 0 && divisor > 0 || dividend < 0 && divisor < 0 {
		positive = true
	}

	if dividend > 0 {
		dividend = -dividend
	}
	if divisor > 0 {
		divisor = -divisor
	}

	_, ans := dfs(dividend, divisor, 1)

	if positive {
		return ans
	} else {
		return -ans
	}
}

func dfs(dividend, divisor int, count int) (int, int) {
	if dividend > divisor {
		return dividend, 0
	}

	d, c := dfs(dividend, divisor+divisor, count+count)
	if d <= divisor {
		return d - divisor, count + c
	}

	return d, c
}

func main() {
	dividend := -2147483648
	divisor := -1
	res := divide(dividend, divisor)

	println(res)
}
